protocols/relay/src/v2/relay/handler.rs

```diff
@@ -0,0 +1,892 @@
+// Copyright 2021 Protocol Labs.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a
+// copy of this software and associated documentation files (the "Software"),
+// to deal in the Software without restriction, including without limitation
+// the rights to use, copy, modify, merge, publish, distribute, sublicense,
+// and/or sell copies of the Software, and to permit persons to whom the
+// Software is furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+// OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+// FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+// DEALINGS IN THE SOFTWARE.
+
+use crate::v2::copy_future::CopyFuture;
+use crate::v2::message_proto::Status;
+use crate::v2::protocol::{inbound_hop, outbound_stop};
+use crate::v2::relay::CircuitId;
+use bytes::Bytes;
+use futures::channel::oneshot::{self, Canceled};
+use futures::future::{BoxFuture, FutureExt, TryFutureExt};
+use futures::io::AsyncWriteExt;
+use futures::stream::{FuturesUnordered, StreamExt};
+use futures_timer::Delay;
+use instant::Instant;
+use libp2p_core::connection::ConnectionId;
+use libp2p_core::either::EitherError;
+use libp2p_core::{upgrade, ConnectedPoint, Multiaddr, PeerId};
+use libp2p_swarm::protocols_handler::{InboundUpgradeSend, OutboundUpgradeSend};
+use libp2p_swarm::{
+    IntoProtocolsHandler, KeepAlive, NegotiatedSubstream, ProtocolsHandler, ProtocolsHandlerEvent,
+    ProtocolsHandlerUpgrErr, SubstreamProtocol,
+};
+use std::collections::VecDeque;
+use std::fmt;
+use std::task::{Context, Poll};
+use std::time::Duration;
+
+pub struct Config {
+    pub reservation_duration: Duration,
```

Comment:
I'd expect there to be a small difference between the duration the server uses before it removes a reservation and the duration a client uses before it renews, but I can't find that difference anywhere. Am I missing it or is it not here?

To elaborate, this is to protect against not network delays and clock skew. If a server says a reservation is good for 1 minute, the client probably wants to renew that at some frequency < 1 min (maybe 50s or 30s to be safe).

Reply:
Great catch!

Somehow I had in my head that I added a hacky constant 10 seconds somewhere. Not the case.

Checking go-libp2p, it uses [a constant 2 mnutes](https://github.com/libp2p/go-libp2p/blob/f7e7cdedf0dff7a7991c2e5bb4b095644d6db6a0/p2p/host/autorelay/autorelay.go#L33) and [a 2h reservation timeout](https://github.com/libp2p/go-libp2p/blob/79d995404cd5b332005d3d9f2c602d7ea3416c15/p2p/protocol/circuitv2/relay/resources.go#L48).

Instead of a constant, or a configuration parameter, I am inclined to just use a percentage of the reservation duration. E.g. if the relay returns 2h, wait for 1h and 30 minutes (75%) before refreshing it.

Thoughts? a60ffbab65f749b5f198cdc0b962c17f3de03f8b would do the corresponding change.

Also maybe @vyzo has opinions on this.

Reply:
3/4ths seems good to me 👍